Allow hostports with localhost hostIPs

aspsk opened the pull request:
We need to enable localhost addresses for hostIP, because this is one of the reasons why the following conformance test fails:

  sonobuoy run --e2e-focus 'validates that there is no conflict between pods with same hostPort but different hostIP and protocol'

And we need all conformance tests to pass.  (This is not the only reason, the other one is that we don't distinguish between TCP and UDP services, yet.)

The patch is pretty straightforward (we previously just didn't let to specify localhost addresses for hostports in the pod initialization routine).  One note, however, is that we need to explicitly check that pods are not sending data to such "localhost" hostports - only the root namespace should be allowed to connect. We need to do this in both bpf_lxc.c and bpf_sock.c for per-connection and per-packet service translations, correspondingly.

```release-note
Allow users to specify hostports with localhost hostIP
```

borkmann commented on bpf/bpf_lxc.c:
```diff
@@ -1258,6 +1262,10 @@ static __always_inline int __tail_handle_ipv4(struct __ctx_buff *ctx)
+
+			if (is_v4_loopback(ip4->daddr))
+				goto skip_service_lookup;
+
```
I think this is not enough given also traffic on the wire could get to the node where we have the 127.0.0.1 as IP in svc table, so also this needs to be rejected from devs with bpf_host/xdp on it.

I wonder.. are hostport ranges forbidden to clash with nodeport ranges? If so, maybe we could solve this issue by installing a wildcard entry, which covers loopback and all node local addresses - similar as we do in kpr for nodeport.

The logic in sock LB, should probably go into sock{4,6}_skip_xlate() which already has checks around lb{4,6}_svc_is_hostport().

aspsk replied:
Thanks @borkmann !

Yeah, martian packets... So, am I right that we are protecting here from hosts which are on the same L2 as our k8s host? Or is there other ways to send DST=127.0.0.1 packets from the outside?

For the port range: I believe that the hostPort range is not defined, so it can be any, but the NodePort range is reserved, so we won't let users to create HostPorts within the NodePort range. As a matter of fact, I've just checked that you added this check in 2020 in a commit named "prep datapath for loopback/local address exposure on hostport", so I will look into installing the wildcard entry :-)

borkmann replied:
> Yeah, martian packets... So, am I right that we are protecting here from hosts which are on the same L2 as our k8s host? Or is there other ways to send DST=127.0.0.1 packets from the outside?

Yes, this exactly, e.g. you could craft af_packet skb and push it to the wire when in same L2.

aspsk committed:
Allow hostports with localhost hostIPs

We need to enable localhost addresses for hostIP, because this is one of the
reasons why the following conformance test fails:

    sonobuoy run --e2e-focus 'validates that there is no conflict between pods with same hostPort but different hostIP and protocol'

And we need all conformance tests to pass.  (This is not the only reason, the
other one is that we don't distinguish between TCP and UDP services, yet.)

In order to implement this, we allow localhost hostports and convert them into
wildcard entries with an additional flag set: SVC_FLAG_LOOPBACK. Then, in
bpp_sock.c we check that connections to such wildcard entries are only resolved
for host namespace sockets, and only to a loopback address. To do this we need
the get_netns_cookie helper to be defined, so this feature is only enabled for
recent (5.7+) kernels.

The loopback hostports have only wildcard entries. They will be marked with the
'loopback' flag in the output of 'cilium bpf lb list', e.g.:

    # cilium bpf lb list
    SERVICE ADDRESS    BACKEND ADDRESS (REVNAT_ID) (SLOT)
    0.0.0.0:12345      0.0.0.0:0 (5) (0) [HostPort, non-routable, loopback]
                       10.244.1.84:1234 (5) (1)
    10.96.0.1:443      0.0.0.0:0 (1) (0) [ClusterIP, non-routable]
                       172.18.0.2:6443 (1) (1)
    10.96.0.10:53      10.244.0.129:53 (2) (2)
                       10.244.0.30:53 (2) (1)
                       0.0.0.0:0 (2) (0) [ClusterIP, non-routable]
    ...

This commit also adds a new bpf unit-test bpf/tests/wildcard_lookup.c which
tests the sock{4,6}_wildcard_lookup{,_full} functions.

Signed-off-by: Anton Protopopov <aspsk@isovalent.com>

joamaki reviewed (approved):
Looks good. General observations: the whole wildcard thing is a bit hard to follow. Maybe the lookup functions could be better documented to show what are the concrete lookups they perform? We should perhaps also consider using bit flags with named constants when we've got more than one boolean flag, e.g. `sock4_wildcard_lookup(key, WCL_INV_MATCH | WCP_IN_HOSTNS)`.

I would recommend adding a test to `bpf/tests` for this as it should not be too difficult now and a good idea to extend our still somewhat poor coverage whenever we touch the bpf code. `host_only_socket_lb_test.c` is probably closest as a starting point.

aspsk committed:
bpf: tests: rename a variable to prevent build failure

Build of a test will fail if a 'key' variable is defined, like this:

    wildcard_lookup.c:123:2: error: declaration shadows a local variable [-Werror,-Wshadow]
            test_init();
            ^
    ./common.h:198:9: note: expanded from macro 'test_init'
                    __u32 key = 0;                                            \
                          ^
    wildcard_lookup.c:113:17: note: previous declaration is here
            struct lb4_key key = {
                           ^
    wildcard_lookup.c:348:2: error: declaration shadows a local variable [-Werror,-Wshadow]
            test_init();
            ^
    ./common.h:198:9: note: expanded from macro 'test_init'
                    __u32 key = 0;                                            \
                          ^
    wildcard_lookup.c:328:17: note: previous declaration is here
            struct lb6_key key = {
                           ^

Fix this by renaming a variable used in test setup macros.

Signed-off-by: Anton Protopopov <aspsk@isovalent.com>

joamaki commented on pkg/k8s/watchers/pod.go:
```diff
@@ -479,6 +483,11 @@ func (k *K8sWatcher) deleteK8sPodV1(pod *slim_corev1.Pod) error {
 }
 
+func netnsCookieSupported() bool {
+	return probes.HaveProgramHelper(ebpf.CGroupSock, asm.FnGetNetnsCookie) == nil &&
```
This will crash tests as they cannot perform this syscall (they're not running as root, or the tests are run under macOS). This should be resolved at daemon initialization and provided via e.g. `option.Config` or somehow passed to k8s watcher. Or can we actually just drop the whole `feIP` check completely? If it would only affect old kernels I don't think there's much value to it if we need to go out of our way to supply the probe result to this method.

The real problem here is that genServiceMappings really doesn't belong in pkg/k8s/watchers. It's something that should be somewhere under pkg/datapath. That's unfortunately bigger refactoring.

aspsk replied:
Tests shoudn't break, as we don't test this functionality from user space

brb replied:
> The real problem here is that genServiceMappings really doesn't belong in pkg/k8s/watchers.

ACK. Just that, unfortunately, at the moment I don't see any way around than extending the refactoring backlog.

brb commented on pkg/k8s/watchers/pod.go:
```diff
@@ -479,6 +483,11 @@ func (k *K8sWatcher) deleteK8sPodV1(pod *slim_corev1.Pod) error {
 }
 
+func netnsCookieSupported() bool {
+	return probes.HaveProgramHelper(ebpf.CGroupSock, asm.FnGetNetnsCookie) == nil &&
```
Shouldn't this be wrapped into `sync.Once()`?

joamaki replied:
The probing result is memoized (in cilium/ebpf's HaveProgramHelper), so sync.Once isn't really needed. This is probably a fairly recent change.

aspsk replied:
Let me keep it @joamaki, I've never used `sync.Once` before :pray: :rofl:

aspsk replied:
@joamaki Actually, I think this is really ok to leave the `sync.Once` here. Even though the `HaveProgramHelper` (more precisely the `ebpf.features.haveProgramHelper`) caches the result, this should be some cycles faster with the `sync.Once`. So I will mark as resolved, feel free to, revert

brb commented on bpf/bpf_sock.c:
```diff
@@ -226,20 +226,29 @@ static __always_inline struct lb4_service *
+	 */
+	key->address = orig_addr;
+	svc = sock4_wildcard_lookup(key, false, true, in_hostns);
+	if (svc && lb4_svc_is_hostport(svc) && (!lb4_svc_is_loopback(svc) || loopback))
```
Why don't we use the following condition instead?

`(svc && lb4_svc_is_hostport(svc) && lb4_svc_is_loopback(svc) && loopback)`

aspsk replied:
@brb This is a different condition ^ it will reject all lookups for non-`SVC_FLAG_LOOPBACK` services. Does the comment in the code above makes sense to you?

brb replied:
Yep, it does. Thanks.

brb reviewed (commented):
LGTM, just a few questions / comments.

brb reviewed (approved):
:rocket:
